src/main/groovy/com/github/jengelman/gradle/plugins/shadow/internal/UnusedTrackerWithR8.groovy

```diff
@@ -0,0 +1,276 @@
+package com.github.jengelman.gradle.plugins.shadow.internal
+
+import com.android.tools.r8.*
+import com.android.tools.r8.origin.Origin
+import groovy.transform.CompileStatic
+import org.apache.commons.io.FilenameUtils
+import org.gradle.api.Project
+import org.gradle.api.artifacts.Configuration
+import org.gradle.api.file.FileCollection
+import org.gradle.api.tasks.SourceSet
+
+import java.nio.file.Files
+import java.nio.file.Path
+import java.nio.file.Paths
+
+/**
+ * An implementation of an UnusedTracker using R8.
+ * <p>
+ * This class is written mainly in Java syntax and uses @CompileStatic deliberately
+ * to avoid groovy madness especially in combination with R8.
+ * <p>
+ * It is an extension of the existing UnusedTracker to make it easy to exchange them,
+ * if the pull request is accepted, a common base interface should be extracted.
+ */
+class UnusedTrackerWithR8 extends UnusedTracker {
```

Comment:
We could just write this in pure Java in the `src/main/java` directory if this is a concern.